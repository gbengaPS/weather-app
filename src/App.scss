body {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  font-size: 1rem;
}

#root {
  height: 100vh;
  width: 100vw;
  display: flex;
  justify-content: center;
}

.app {
  width: 100%;
  max-width: 960px;
  padding: 20px;
}

.city-list {
  margin: 0;
  padding: 0;
  
  li {
    width: 100%;
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin: 0 0 5px 0;
    background-color: #5bb3ff;
    border-radius: 5px;
    padding: 5px;
  }

  li:hover {
    box-shadow: #fff;
  }
}

.city-info {
  display: flex;
  justify-content: space-between;
  align-items: center;
  width: 80%;

  img {
    width: 50px;
    height: 50px;
  }
}
.temperature {
  font-weight: 700;
  font-size: 1.4rem;
  color: #fff;
}

.aligned-flex {
  display: flex;
  justify-content: space-between;
  align-items: center;
}