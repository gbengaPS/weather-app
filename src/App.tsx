import React, { useEffect, useContext } from 'react';
import { Link, useHistory } from 'react-router-dom';
import LargestCities from './components/LargestCities';
import Search from './components/Search';
import {
  getCityFromAPI,
  usePopulateStore,
  useOfflineIndicator,
  getUserCurrentCity
} from './helpers';
import { AppContext } from './store';
import { ADD_FAVORITE } from './constants';
import './styles/app.scss';

const App = () => {
  const { state, dispatch } = useContext(AppContext);
  const history = useHistory();

  usePopulateStore();
  useOfflineIndicator();

  useEffect(() => {
    navigator.geolocation.getCurrentPosition(async (data) => {
      if (!localStorage.getItem('hasSeenCurrentCity')) {
        const { coords } = data;
        const response = await getCityFromAPI({ lat: coords.latitude, lon: coords.longitude });
        
        if (response) {
          localStorage.setItem('userCurrentCity', JSON.stringify(response));
          localStorage.setItem('hasSeenCurrentCity', 'yes');
          history.push('user/location');
        }
      }
    });
  }, [history]);

  useEffect(() => {
    const favorites = JSON.parse(localStorage.getItem('favorites') || '{}');
    dispatch({ type: ADD_FAVORITE, favoriteItem: favorites })
  }, [dispatch])

  if (state.isLoading) {
    return <img src="/loader.gif"  alt="loading indicator"/>
  }

  const favoritesArray = Object.entries(state.favorites).sort((first, second) => {
    return first[1].data.name.localeCompare(second[1].data.name)
  }).map((entry) => entry[1]);

  const sortedWeatherInfo = state.weatherInfo.sort((first, second) => {
    return first.data.name.localeCompare(second.data.name);
  })

  const largestCitiesInfo = [...favoritesArray, ...sortedWeatherInfo];

  const userCurrentCity = getUserCurrentCity();

  return (
    <div className="app">
      {
        userCurrentCity &&
        <p className="no-margin"><Link to="/user/location">My location</Link></p>
      }
     
    
      <Search />
      <p><i className="fas fa-long-arrow-alt-left"></i></p>
      <LargestCities
        cities={largestCitiesInfo}
      />
    </div>
  );
}

export default App;
