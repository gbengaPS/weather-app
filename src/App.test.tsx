import React from 'react';
import { render } from '@testing-library/react';
import {  BrowserRouter as Router, Route } from 'react-router-dom';
import App from './App';

Object.defineProperty(window.navigator, 'geolocation',
{ value: { getCurrentPosition: jest.fn()}, writable: true });

describe('App', () => {
  beforeEach(() => {

  })
  it('should match snapshot', () => {
    const { container } = render(
      <Router>
        <Route>
          <App />
        </Route>
      </Router>
    );
    expect(container).toMatchSnapshot();
  })
})
