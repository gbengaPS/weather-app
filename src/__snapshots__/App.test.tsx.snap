// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div>
  <div
    class="app"
  >
    <a
      href="/user/location"
    >
      My location
    </a>
    <div
      class="search-wrapper"
    >
      <form>
        <input
          aria-label="Search city weather information"
          id="cities-search"
          placeholder="Search cities"
          type="search"
          value=""
        />
        <button
          data-testid="search-button"
        >
          <i
            aria-hidden="true"
            class="fa fa-search"
          />
        </button>
      </form>
    </div>
    <div
      class="search-result-wrapper"
    >
      
    </div>
    <p>
      <i
        class="fas fa-long-arrow-alt-left"
      />
    </p>
    <p
      class="error"
    >
      No weather information to display.
    </p>
  </div>
</div>
`;
